File: LeetCodeRegular/intersection-of-two-arrays-ii.cpp
class Solution {
public:
    vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size(), m = nums2.size();

        sort(nums1.begin(), nums1.end()); //可以用标准库直接对数组进行排序
        sort(nums2.begin(), nums2.end());

        int i = 0, j = 0;
        vector<int> interSection;

        while( i < n && j < m){
            if(nums1[i] < nums2[j]){
                i++;
            }
            else if(nums1[i] > nums2[j]){
                j++;
            }
            else{
                interSection.push_back(nums1[i]); //给向量插入值用push_back，不能直接赋值
                i++;
                j++;
            }
        }
        return interSection;
        
    }
};

/*
如果两个数组是有序的，则可以便捷地计算两个数组的交集。

首先对两个数组进行排序，然后使用两个指针遍历两个数组。

初始时，两个指针分别指向两个数组的头部。每次比较两个指针指向的两个数组中的数字，如果两个数字不相等，则将指向较小数字的指针右移一位，如果两个数字相等，
将该数字添加到答案，并将两个指针都右移一位。当至少有一个指针超出数组范围时，遍历结束。
（这是我的解题思路，与官方题解第二个思路一致，因此直接将官方解答放在这里）
*/

File: LeetCodeRegular/数据结构/数据结构入门/数组/LC_00350/intersection-of-two-arrays-ii.py
'''
350. 两个数组的交集 II
给你两个整数数组 nums1 和 nums2 ，请你以数组形式返回两数组的交集。返回结果中每个元素出现的次数，
应与元素在两个数组中都出现的次数一致（如果出现次数不一致，则考虑取较小值）。可以不考虑输出结果的顺序。
'''

class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        m = len(nums1)
        n = len(nums2)

        nums1.sort()
        nums2.sort()
        
        ans = []
        i,j = 0, 0 

        while i < m and j < n:
            if nums1[i] < nums2[j]:
                i += 1
            elif nums1[i] > nums2[j]:
                j += 1
            else:
                ans.append(nums1[i])
                i += 1
                j += 1
        
        return ans

 #排序+双指针 
